sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala

```diff
@@ -218,24 +218,20 @@ private[sql] case class InsertIntoHadoopFsRelation(
         val partitionProj = newProjection(codegenEnabled, partitionOutput, output)
         val dataProj = newProjection(codegenEnabled, dataOutput, output)
 
-        if (needsConversion) {
-          val converter = CatalystTypeConverters.createToScalaConverter(dataSchema)
-          while (iterator.hasNext) {
-            val row = iterator.next()
-            val partitionPart = partitionProj(row)
-            val dataPart = dataProj(row)
-            val convertedDataPart = converter(dataPart).asInstanceOf[Row]
-            writerContainer.outputWriterForRow(partitionPart).write(convertedDataPart)
-          }
+        val partitionSchema = StructType.fromAttributes(partitionOutput)
+        val partConverter = CatalystTypeConverters.createToScalaConverter(partitionSchema)
```

Comment:
Hmm, looks like I got confused here.  I thought that we were converting to Catalyst, but I guess we're actually converting from Catalyst to Scala.  In that case, I guess it makes sense that we need to convert the result of the projection in the `!needsConversion` branch, since even if we already have a Scala row its projection might be Catalyst.  It looks like the old code might have actually contained a mistake here, since I think we'd also need to convert the projection in the `needsConversion` branch, too.  Therefore, this looks okay to me.